flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultResultPartition.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adapter;
+
+import org.apache.flink.runtime.executiongraph.IntermediateResultPartition;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingExecutionVertex;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.DONE;
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.EMPTY;
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.PRODUCING;
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.RELEASED;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Default implementation of {@link SchedulingResultPartition}.
+ */
+public class DefaultResultPartition implements SchedulingResultPartition {
+
+	private final IntermediateResultPartition resultPartition;
+
+	private final SchedulingExecutionVertex producer;
+
+	private final List<SchedulingExecutionVertex> consumers;
+
+	DefaultResultPartition(IntermediateResultPartition partition, SchedulingExecutionVertex producer) {
+		this.resultPartition = checkNotNull(partition);
+		this.producer = checkNotNull(producer);
+		this.consumers = new ArrayList<>();
+	}
+
+	@Override
+	public IntermediateResultPartitionID getId() {
+		return resultPartition.getPartitionId();
+	}
+
+	@Override
+	public IntermediateDataSetID getResultId() {
+		return resultPartition.getIntermediateResult().getId();
+	}
+
+	@Override
+	public ResultPartitionType getPartitionType() {
+		return resultPartition.getResultType();
+	}
+
+	@Override
+	public ResultPartitionState getState() {
+		switch (producer.getState()) {
+			case CREATED:
+			case SCHEDULED:
+			case DEPLOYING:
+				return EMPTY;
+			case RUNNING:
+				return PRODUCING;
+			case FINISHED:
+				return DONE;
+			default:
+				return RELEASED;
+		}
+	}
+
+	@Override
+	public SchedulingExecutionVertex getProducer() {
+		return producer;
+	}
+
+	@Override
+	public Collection<SchedulingExecutionVertex> getConsumers() {
+		return Collections.unmodifiableCollection(consumers);
+	}
+
+	void addConsumer(SchedulingExecutionVertex vertex) {
```

Comment:
Can't we assign the `consumers` in the constructor?

Reply:
In current implementation, I would like to initialize `DefaultResultPartition` by the producedPartitions when initializing `DefaultExecutionVertex`, and the consumer `DefaultExecutionVertex` has not been initialized, thus not easy to assign the `consumers` in the constructor.

Another consideration of using `addConsumer` is that I would like to initialize the information of `DefaultExecutionVertex#consumedPartitions` and `DefaultResultPartition#consumers` in one step to avoid creating contradicting topologies.

Reply:
I'm wondering whether this circular dependency between the partitions and vertices isn't a design flaw in the interfaces. (Also applies to the Failover interfaces)

Similarly to the existing `Execution*` classes, we create initially independent nodes and then retroactively bake a graph structure into them by setting partitions/producers respectively.

Maybe we could instead have the vertices only be these really independent nodes (with an ID and state), setup the partitions afterwards as the graph structure (which, now that I think about it, are awfully close to `ExecutionEdges`, the differentiation becomes quite blurry), and use the vertices to lookup partitions in the graph. (something like `SchedulingTopology#getProcudedResultPartitions(ExecutionVertexID id)`)

@tillrohrmann @GJL WDYT?

Reply:
I think you are right that in order to break up the cyclic dependency we would need to decouple the `SchedulingResultPartition` from the `SchedulingExecutionVertex` (via the `ExecutionVertexID` for example).

I'm wondering how bad this cyclic dependency and the need for mutable state is, though. From a user's perspective, the existing interfaces are a bit easier and more convenient to use. On the down side, it makes the implementation a bit harder and harder to test in isolation. However, do we want to test the `Scheduling*` implementations in isolation? Moreover, you always have this problem in graph structures which have bidirectional edges or loops.